### PR TITLE
Direct links to narrative sections

### LINKED ISSUE #395: Allow direct links to narrative sections
**What needs to be done? Why does it need to be done?**
v2.0 was supposed to launch with this feature, but it proved too buggy and was backed out of the launch. The original scope included: 

1. the sections snap into place when scrolling or using the navigation buttons
1. the page URL updates as you move up and down the page
1. navigating directly to a URL that specifies a section should scroll that section into view

The initial approach made one long scrolling page and used IntersectionObservers to react to sections coming into view. That worked OK for 2 but completely fell over on 3. The changing height of the page as items loaded or changed size caused that failure and a host of other bugs (e.g. resizing the window made things jump around). The solution for 1 was also mostly ineffective; the snapping behavior was inconsistent from browser to browser due to challenges with distinguishing "inertial" wheel events from real ones.

These issues should be solved and the feature restored.

**Additional context**
Two alternatives have been considered: 

1. Keep squashing bugs until we're happy with the behavior.
1. Reconfigure the page to behave more like a slideshow than a long page.

1 is not very appealing because the existing behavior is complex and brittle and piling on more special-case handling to squash the bugs is likely to make it more so. If a more fundamental and elegant solution could be found, that would be better, but that doesn't seem likely. 

2 feels conceptually cleaner (less swimming upstream against native browser behavior) but has more unknowns. It has been explored some but not yet reached the proof-of-concept stage. It also potentially involves more "scroll-jacking" behavior which might be a performance concern. A research spike would be advised before committing to this approach.

### PR DESCRIPTION
**Pull request by macfarlandian**
## Description of the change

Revives and revamps the direct section linking feature (disabled [here](https://github.com/Recidiviz/public-dashboard/pull/375/commits/d3c4f270e03d01ec4343597b64b516c927dcdf25) to unblock the v2 launch). 

(The new behavior is different enough from the old that it's probably not a useful reference, but I include it nonetheless for context).

In some ways I think this is less complicated than it was the first time around because it relies on native browser behavior rather than JS for smooth scrolling and snapping. But it's not _not_ complicated ... the new complexity here involves keeping track of where the user is on the page to prevent the linked section from being pushed down the viewport by content loading above it. 

The end result should be that you can link directly to any section and it will scroll cleanly into your viewport when the page loads, and the page URL will stay in sync as you move up and down the page and can be grabbed at any time from the location bar to reflect your current view.

To summarize the highlights of how we got there: 

- On direct visits to a URL specifying a section > 1, the desired section is scrolled into view and any sections above it have their heights clamped to the height of the viewport; this prevents them from pushing the linked section out of position. When these sections scroll into view (i.e. if the user moves up the page), they expand to their natural height as needed. This expansion is animated to prevent any jarring "jumps" up the page; it's most noticeable on mobile where the sections are several screens tall.
- On direct visits to the top of the page (no section or section 1), **none** of the above happens.
- We detect when the top or bottom of an adjacent section comes into view and update the page URL to match. This responds to both scrolling and navigation via the left side buttons. (We were already doing this but I have improved the sensitivity somewhat.)
- Sections will snap to the top of page when they are nearby, using the native [Scroll Snap API](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Scroll_Snap/Basic_concepts), which was new to me but works pretty well, though a little differently from browser to browser. It's supported everywhere but IE 11 so is a good candidate for progressive enhancement, since it's not critical. This lets us drop the previous JS scrolling-and-snapping solution entirely, which is a net gain because that code pretty much didn't work at all in any browser 😎. The caveat here is that it gets confused by the apparent "scrolling" that occurs when the aforementioned clamped sections expand, and so needs to be disabled until we're done expanding all the sections. Again ... progressive enhancement 🎉 .

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

> Closes #395 

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing against realistic data has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: spotlight-client/src/NarrativeLayout/NarrativeLayout.tsx [hunk 1]
```diff
@@ -55,108 +56,62 @@ type NarrativeLayoutProps = {
...
-  }, [activeSection, sectionsContainerRef, setScrollSpring]);
+  const {
+    alwaysExpanded,
+    currentSectionNumber,
+    enableSnapping,
+    fixedHeightSections,
+    scrollToSection,
+    sectionsContainerRef,
+    getOnSectionExpanded,
+    onInViewChange,
```
Comment (daschi) [inline spotlight-client/src/NarrativeLayout/NarrativeLayout.tsx:69]: Wondering what you think about calling this variable `onSectionInViewChange`, might be easier to read later on without needing to checkout the note in the hook.

**Review by daschi — COMMENTED** [review]
This looks awesome...I noticed a behavior that I'm not sure is intended:

1. Navigate to a section by entering it into the url (i.e. prison/3)
2. Select a different Data Narrative (i.e. probation)
3. See that it navigates to probation/3 - I would have expected it to go to probation/1

**Comment by macfarlandian** [comment]
@daschi hmmm good catch, that is indeed not intended. i'll look into it!

**Comment by macfarlandian** [comment]
turns out the same layout component was re-rendering when the narrative changed so the "initial" section of the first load was persisting as you navigated. (this wasn't happening until I consolidated everything into a hook instead of passing some of it as props, which was the last thing I did before opening this PR, so my earlier testing didn't catch it.) I added a more targeted reaction to clear it when you go to a new page, using a local Mobx observable

**Review by daschi — APPROVED** [review]

**Pull request merged**